maven_iuvs/echelle.py

```diff
@@ -334,11 +334,163 @@ def downselect_data(index, light_dark=None, orbit=None, date=None, segment=None,
 
 # Relating to dark vs. light observations -----------------------------
 
-def update_master_lightdark_key(key_filename, ech_l1a_idx, dark_idx, 
+def update_filenames_in_light_dark_key(keyfile, ech_l1a_idx, dark_idx, 
+                                       verbose=True):
+    """
+    Searches the light/dark key csv (keyfile) for any filenames that have since 
+    had a revision update. In that sense, the "copy of record" is the copy on 
+    the user's local hard drive. 
+
+    Parameters
+    ----------
+    keyfile : string
+              .csv file containing columns 'Light Folder', 'Light', 'Dark  
+              Folder', 'Dark', 'DTobj, 'Segment'.
+    ech_l1a_idx : list of dictionaries
+                  Metadata of local files, used to determine newest filename.
+    dark_idx : list of dictionaries
+               Same as above but the dark files only.
+    verbose : bool
+              If True will print result messages.
+
+    Returns
+    ----------
+    MASTER_KEY : pandas DataFrame
+                 New master key. 10s of thouasnds of lines
+    changes : pandas DataFrame
+              Lists old files and what they were replaced with
+    problems : pandas DataFrame
+               Files that caused a problem
+    """
+    MASTER_KEY = pd.read_csv(keyfile)
+
+    changes = {"row": [], "oldname": [], "newname": []}
+    problems = {"row": [], "oldname": [], "newname": []}
+
+    # Precompile regex
+    versioned_pat = re.compile(r"^(?P<base>mvn.+?)_(r|s)(?P<rev>\d+)(?P<ext>\.fits\.gz)$")
+
+    # Cache for folder listings
+    light_folder_cache = {}
+    dark_folder_cache = {}
+
+    new_lightnames = MASTER_KEY["Light"].copy()
+    new_darknames = MASTER_KEY["Dark"].copy()
+
+    for i, row in tqdm(MASTER_KEY.iterrows(), total=len(MASTER_KEY)):
+
+        # Lights
+        lfolder = row["Light Folder"]
+        dfolder = row["Dark Folder"]
+        original_lightname = row["Light"]
+        original_darkname = row["Dark"]
+
+        # Cache file list in this folder
+        if lfolder not in light_folder_cache:
+            light_folder_cache[lfolder] = os.listdir(lfolder)
+
+        # Extract base name and extension
+        lmatch = versioned_pat.match(original_lightname)
+        if not lmatch:
+            problems["row"].append(i)
+            problems["oldname"].append(original_lightname)
+            problems["newname"].append("Regex match failed")
+            continue
+
+        base = lmatch.group("base")
+        rev = int(lmatch.group("rev"))
+        ext = lmatch.group("ext")
+
+        # Find all matching versions in the folder
+        lmatches = []
+        for f in light_folder_cache[lfolder]:
+            m = versioned_pat.match(f)
+            if m:
+                f_base = m.group("base")
+                f_rev = int(m.group("rev"))
```

Comment:
how do you handle stage vs. production files?

Reply:
Well, it was initially treating the copy on disk as the the copy to prefer. But I think that could backfire in rare circumstances. So I rebuilt the whole thing to make it more efficient, compare r vs. s and numbers of revision, and not duplicate code. 43aca59752dfa8dcfe3b41ed84d20b0d3feb7087

Reply:
It now compares r and s, preferring r, and then just picks the biggest revision number available. You may want to have a quick look at the update. I only check to see if the disk file is newer in one of those ways - if the file listed in the light/dark key csv is newer than whatever the user has on disk, nothing is done. (but I don't know how that would be different anyway. It would be weird to ask that routine to go download new files!)

Reply:
I still feel like this should be able to rely on `get_latest_files` somehow, but it's OK to leave this functionality as separate for now.